File: Backend/src/main/java/com/shopsphere/backend/service/ProductService.java
package com.shopsphere.backend.service;

import com.shopsphere.backend.dto.product.ProductRequest;
import com.shopsphere.backend.dto.product.ProductResponse;

import java.math.BigDecimal;
import java.util.List;

public interface ProductService {

    // ==================================================
    // CREATE
    // ==================================================

    ProductResponse createProduct(
            ProductRequest request
    );

    // ==================================================
    // READ
    // ==================================================

    List<ProductResponse> getAllProducts();

    ProductResponse getProductById(
            Long id
    );

    // ==================================================
    // PAGINATION
    // ==================================================

    List<ProductResponse> getProductsWithPagination(
            int page,
            int size
    );

    // ==================================================
    // UPDATE
    // ==================================================

    ProductResponse updateProduct(
            Long id,
            ProductRequest request
    );

    // ==================================================
    // ADD STOCK
    // ==================================================

    ProductResponse addStock(
            Long id,
            Integer quantity
    );

    // ==================================================
    // DELETE
    // ==================================================

    void deleteProduct(
            Long id
    );

    // ==================================================
    // SEARCH
    // ==================================================

    List<ProductResponse> searchProducts(
            String keyword
    );

    // ==================================================
    // CATEGORY FILTER
    // ==================================================

    List<ProductResponse> getProductsByCategory(
            Long categoryId
    );

    // ==================================================
    // PRICE FILTER
    // ==================================================

    List<ProductResponse> filterByPrice(
            BigDecimal minPrice,
            BigDecimal maxPrice
    );

    // ==================================================
    // SORTING
    // ==================================================

    List<ProductResponse> sortByPriceAsc();

    List<ProductResponse> sortByPriceDesc();

    List<ProductResponse> getLatestProducts();

    // ==================================================
    // COMBINED SEARCH, FILTER AND SORT
    // ==================================================

    List<ProductResponse> filterProducts(
            String keyword,
            Long categoryId,
            BigDecimal minPrice,
            BigDecimal maxPrice,
            String sort
    );
}

File: shopsphere-frontend/src/services/productService.js
import axiosInstance from "../api/axios";

// Get all products
export const getAllProducts = () => {
  return axiosInstance.get("/products");
};

// Get one product by ID
export const getProductById = (id) => {
  return axiosInstance.get(
    `/products/${id}`
  );
};

// Search products
export const searchProducts = (
  keyword
) => {
  return axiosInstance.get(
    "/products/search",
    {
      params: {
        keyword,
      },
    }
  );
};

// Get products by category
export const getProductsByCategory = (
  categoryId
) => {
  return axiosInstance.get(
    `/products/category/${categoryId}`
  );
};

// Filter products by price
export const filterProductsByPrice = (
  minPrice,
  maxPrice
) => {
  return axiosInstance.get(
    "/products/filter",
    {
      params: {
        minPrice,
        maxPrice,
      },
    }
  );
};

// Sort products by price:
// Low to high
export const sortProductsByPriceAsc =
  () => {
    return axiosInstance.get(
      "/products/sort/price-asc"
    );
  };

// Sort products by price:
// High to low
export const sortProductsByPriceDesc =
  () => {
    return axiosInstance.get(
      "/products/sort/price-desc"
    );
  };

// Get latest products
export const getLatestProducts = () => {
  return axiosInstance.get(
    "/products/latest"
  );
};

// Combined search, filter,
// and sorting API
export const filterProducts = ({
  keyword,
  categoryId,
  minPrice,
  maxPrice,
  sort,
}) => {
  const params = {};

  if (keyword?.trim()) {
    params.keyword =
      keyword.trim();
  }

  if (
    categoryId !== "" &&
    categoryId !== null &&
    categoryId !== undefined
  ) {
    params.categoryId =
      categoryId;
  }

  if (
    minPrice !== "" &&
    minPrice !== null &&
    minPrice !== undefined
  ) {
    params.minPrice =
      minPrice;
  }

  if (
    maxPrice !== "" &&
    maxPrice !== null &&
    maxPrice !== undefined
  ) {
    params.maxPrice =
      maxPrice;
  }

  if (sort) {
    params.sort = sort;
  }

  return axiosInstance.get(
    "/products/filter-products",
    {
      params,
    }
  );
};

// ==========================================
// ADMIN: ADD PRODUCT STOCK
// ==========================================

export const addProductStock = (
  productId,
  quantity
) => {
  return axiosInstance.put(
    `/products/${productId}/stock`,
    null,
    {
      params: {
        quantity,
      },
    }
  );
};